shotover-proxy/src/transforms/mod.rs

```diff
@@ -521,6 +582,64 @@ pub trait Transform: Send {
     /// Wrapping it in an [`Arc<Mutex<_>>`](std::sync::Mutex) would make it a global count of all messages seen by this transform.
     async fn transform<'a>(&'a mut self, message_wrapper: Wrapper<'a>) -> ChainResponse;
 
+    /// This method should be should be implemented by your transform if it is required to process pushed messages (typically events
+    /// or messages that your source is subscribed to. The wrapper object contains the queries/frames
+    /// in a [`Vec<Message`](crate::message::Message).
+    ///
+    /// This transform method is not the same request/response model as the other transform method.
+    /// This method processes one pushed message before sending it in reverse on the chain back to the source.
+    ///
+    /// You can modify the messages in the wrapper struct to achieve your own designs. Your transform can
+    /// also modify the response from `message_wrapper.call_next_transform_rev` if it needs to. As long as the message
+    /// carries on through the chain, it will function correctly. You are able to add or remove messages as this method is not expecting
+    /// request/response pairs.
+    ///
+    /// ## Invariants
+    /// * _Non-terminating_ - Your `transform_rev` method should not be terminating as the messages should get passed back to the source, where they will terminate.
+    ///
+    ///
+    /// A basic reverse transform that logs event data and counts the number of events it sees could be defined like so:
+    /// ```
+    /// use shotover_proxy::transforms::{Transform, Wrapper};
+    /// use async_trait::async_trait;
+    /// use tracing::info;
+    /// use shotover_proxy::error::ChainResponse;
+    ///
+    /// #[derive(Debug, Clone)]
+    /// pub struct Printer {
+    ///     counter: i32,
+    /// }
+    ///
+    /// impl Default for Printer {
+    ///     fn default() -> Self {
+    ///         Self::new()
+    ///     }
+    /// }
+    ///
+    /// impl Printer {
+    ///     pub fn new() -> Printer {
+    ///         Printer { counter: 0 }
+    ///     }
+    /// }
+    ///
+    /// #[async_trait]
+    /// impl Transform for Printer {
+    ///     async fn transform<'a>(&'a mut self, message_wrapper: Wrapper<'a>) -> ChainResponse {
+    ///         self.counter += 1;
+    ///         info!("{} Event content: {:?}", self.counter, message_wrapper.messages);
+    ///         let response = message_wrapper.call_next_transform_rev().await;
+    ///         response
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// In this example `counter` will contain the count of the number of messages seen for this connection.
+    /// Wrapping it in an [`Arc<Mutex<_>>`](std::sync::Mutex) would make it a global count of all messages seen by this transform.
+    async fn transform_rev<'a>(&'a mut self, message_wrapper: Wrapper<'a>) -> ChainResponse {
```

Comment:
I wonder if we should call this function `transform_event`?
While it certainly does go in reverse, rev doesnt explain the type of messages that it receives.
Like it might sound like all message replies come through this transform.
Whereas `transform_event` still implies that the pipeline goes in reverse because events go from the server to the client.
Maybe even `transform_server_event` would make it more obvious that its an event coming from the server?

`event` refers to the type of messages that cassandra sends down this pipeline.
The equivalent for redis would be `publish` messages, that doesnt quite explain whats going on as well, so I think `event` is the better general name.

Reply:
Perhaps `pushed`? I think that describes both well.

Reply:
Does pushed imply a direction?
I think a client could be considered pushing to a server?
`event` implies coming from the server which is useful property.
But maybe I only think that because of my knowledge of cassandra...

Reply:
"Pushed" as in, the messages are not a response to a request, but rather pushed from the server to the client. Think "push notifications" on a mobile phone.

Reply:
Right true, I agree it is natural to consider it from the clients perspective so pushed sounds like it is being pushed from the server.
pushed sounds reasonable to me :+1: